### PR TITLE
feat: add priority tax reactor

### PR DESCRIPTION
**Pull request by zhongeric**
Introduces a new reactor type, PriorityOrderReactor, which scales the output amount required to fill an order by the priority fee set in the filling transaction

### CODE REVIEW

File: src/lib/PriorityFeeLib.sol [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    uint256 constant MPS = 1e10;
+
+    /// @notice returns a scaled input using the current priority fee and mpsPerPriorityFeeWei
+    /// @notice the amount is scaled down to favor the swapper
+    /// @param input the input to scale
+    /// @param priorityFee the current priority fee
+    /// @return a scaled input
+    function scale(PriorityInput memory input, uint256 priorityFee) internal pure returns (InputToken memory) {
+        if (priorityFee * input.mpsPerPriorityFeeWei >= MPS) {
+            return InputToken({token: input.token, amount: 0, maxAmount: 0});
```
Comment (marktoda) [inline src/lib/PriorityFeeLib.sol:22]: ok crazy but is there ever a case where the MEV is so high they'll pay YOU?

File: src/lib/PriorityFeeLib.sol [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    /// @notice returns scaled outputs using the current priority fee and mpsPerPriorityFeeWei
+    function scale(PriorityOutput[] memory outputs, uint256 priorityFee)
+        internal
+        pure
+        returns (OutputToken[] memory result)
+    {
+        uint256 outputLength = outputs.length;
+        result = new OutputToken[](outputLength);
+        unchecked {
```
Comment (marktoda) [inline src/lib/PriorityFeeLib.sol:49]: not needed in 0.8.2x

File: src/reactors/PriorityOrderReactor.sol [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    function _resolve(SignedOrder calldata signedOrder)
+        internal
+        view
+        override
+        returns (ResolvedOrder memory resolvedOrder)
+    {
+        PriorityOrder memory priorityOrder = abi.decode(signedOrder.order, (PriorityOrder));
+        _validateOrder(priorityOrder);
+
+        uint256 priorityFee = tx.gasprice - block.basefee;
```
Comment (marktoda) [inline src/reactors/PriorityOrderReactor.sol:34]: one thing to make sure we consider is how different L2s handle these fields - I assume the default handling is the same as mainnet eip-1559 but potentially some handle differently due to separation of L1 data fee, slightly different gas rules etc.
Worth adding a comment to the contract I think clarifying that this only works on chains that have xyz properties

**Review by marktoda — COMMENTED** [review]
very sick - few small thoughts

File: src/lib/PriorityFeeLib.sol (hunk 1 above)
Comment (zhongeric) [reply src/lib/PriorityFeeLib.sol:22]: discussed offline - don't think this can happen since the MEV is naturally capped at the max input offered by the swapper

File: src/reactors/PriorityOrderReactor.sol (hunk 3 above)
Comment (zhongeric) [reply src/reactors/PriorityOrderReactor.sol:34]: oooo good callout, so it's possible that this calculated value could include other additional fees / not represent the actual priority fee correctly?

File: src/lib/PriorityOrderLib.sol [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    function hash(PriorityOrder memory order) internal pure returns (bytes32) {
+        return keccak256(
+            abi.encode(
+                ORDER_TYPE_HASH,
+                order.info.hash(),
+                order.startBlock,
+                order.input.token,
+                order.input.amount,
+                order.input.mpsPerPriorityFeeWei,
+                hash(order.outputs)
```
Comment (snreynolds) [inline src/lib/PriorityOrderLib.sol:102]: order.outputs.hash() syntax?

File: src/reactors/PriorityOrderReactor.sol [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            revert InvalidDeadline();
+        }
+
+        if (order.startBlock > block.number) {
+            revert OrderNotFillable();
+        }
+
+        if (order.input.mpsPerPriorityFeeWei > 0) {
+            for (uint256 i = 0; i < order.outputs.length; i++) {
+                if (order.outputs[i].mpsPerPriorityFeeWei > 0) {
```
Comment (snreynolds) [inline src/reactors/PriorityOrderReactor.sol:75]: oh ok so you can only scale either the input or the output?

Comment (zhongeric) [reply src/reactors/PriorityOrderReactor.sol:75]: Yes only one can be scaled at a time

Comment (zhongeric) [reply src/reactors/PriorityOrderReactor.sol:75]: For exact in / exact output

File: src/lib/PriorityFeeLib.sol [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+    /// @notice we denominate priority fees in terms of milli-bips, or one thousandth of a basis point
+    uint256 constant MPS = 1e7;
+
+    /// @notice returns a scaled input using the current priority fee and mpsPerPriorityFeeWei
+    /// @notice this value is bounded by 0 since the maximal extractable value is <= the full input amount
+    /// @notice the amount is scaled down to favor the swapper
+    /// @notice maxAmount is set to be the original amount and is used to rebuild the permit2 token permissions struct
+    /// @param input the input to scale
+    /// @param priorityFee the current priority fee
```
[the remark this answers is not in the document]
Comment (zhongeric) [reply src/lib/PriorityFeeLib.sol:20]: priority fee should be in wei

File: src/lib/PriorityFeeLib.sol [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+import {OutputToken, InputToken} from "../base/ReactorStructs.sol";
+import {PriorityOutput, PriorityInput} from "../lib/PriorityOrderLib.sol";
+import {FixedPointMathLib} from "solmate/src/utils/FixedPointMathLib.sol";
+
+/// @notice helpers for handling priority order objects
+library PriorityFeeLib {
+    using FixedPointMathLib for uint256;
+
+    /// @notice we denominate priority fees in terms of milli-bips, or one thousandth of a basis point
```
[the remark this answers is not in the document]
Comment (zhongeric) [reply src/lib/PriorityFeeLib.sol:12]: here's some reasoning:
- want to use relative scaling (increase by X bips per wei of priority fee)
- can't use bips as the denominator here because does not give fillers enough points to fill at on the curve (30 bps of price improvement = max bid of 30 wei)
- so we pick something reasonably smaller than bips but not too small where too much is being paid to the sequencer

File: src/lib/PriorityFeeLib.sol [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    /// @notice the amount is scaled down to favor the swapper
+    /// @notice maxAmount is set to be the original amount and is used to rebuild the permit2 token permissions struct
+    /// @param input the input to scale
+    /// @param priorityFee the current priority fee
+    /// @return a scaled input
+    function scale(PriorityInput memory input, uint256 priorityFee) internal pure returns (InputToken memory) {
+        if (priorityFee * input.mpsPerPriorityFeeWei >= MPS) {
+            return InputToken({token: input.token, amount: 0, maxAmount: input.amount});
+        }
+        uint256 scaledAmount = input.amount.mulDivDown((MPS - (priorityFee * input.mpsPerPriorityFeeWei)), MPS);
```
[the remark this answers is not in the document]
Comment (zhongeric) [reply src/lib/PriorityFeeLib.sol:26]: it's not guaranteed, mpsPerPriorityFeeWei * priorityFee is the number of mps to increase the outputs by. If that is >= MPS (total possible), then we set it equal to 0. This is the equivalent of saying that you want to reduce something by 100%.

File: src/lib/PriorityOrderLib.sol [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    address recipient;
+}
+
+/// @dev External struct used to specify priority orders
+struct PriorityOrder {
+    // generic order information
+    OrderInfo info;
+    // the block at which the order becomes active
+    uint256 startBlock;
+    // The tokens that the swapper will provide when settling the order
```
[the remark this answers is not in the document]
Comment (zhongeric) [reply src/lib/PriorityOrderLib.sol:29]: we need to put the scaling factor on both of them to denote if the order is exact in or exact out

File: src/lib/PriorityFeeLib.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (snreynolds) [reply src/lib/PriorityFeeLib.sol:26]: Wouldnt that scenario cause an underflow here then? Cause we're doing 
`(MPS - (priorityFee * input.mpsPerPriorityFeeWei)`

[the remark this answers is not in the document]
Comment (snreynolds) [reply src/lib/PriorityFeeLib.sol:26]: oh i see its checked above! didn't catch that

File: src/lib/PriorityFeeLib.sol [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+import {OutputToken, InputToken} from "../base/ReactorStructs.sol";
+import {PriorityOutput, PriorityInput} from "../lib/PriorityOrderLib.sol";
+import {FixedPointMathLib} from "solmate/src/utils/FixedPointMathLib.sol";
+
+/// @notice helpers for handling priority order objects
+library PriorityFeeLib {
+    using FixedPointMathLib for uint256;
+
+    /// @notice we denominate priority fees in terms of milli-bips, or one thousandth of a basis point
+    uint256 constant MPS = 1e7;
```
Comment (marktoda) [inline src/lib/PriorityFeeLib.sol:13]: Consider changing to pips after our chat w/ Dan?

File: src/lib/PriorityFeeLib.sol [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+    /// @notice returns a scaled input using the current priority fee and mpsPerPriorityFeeWei
+    /// @notice this value is bounded by 0 since the maximal extractable value is <= the full input amount
+    /// @notice the amount is scaled down to favor the swapper
+    /// @notice maxAmount is set to be the original amount and is used to rebuild the permit2 token permissions struct
+    /// @param input the input to scale
+    /// @param priorityFee the current priority fee
+    /// @return a scaled input
+    function scale(PriorityInput memory input, uint256 priorityFee) internal pure returns (InputToken memory) {
+        if (priorityFee * input.mpsPerPriorityFeeWei >= MPS) {
```
Comment (marktoda) [inline src/lib/PriorityFeeLib.sol:23]: nit: cache `priorityfee * input.mpsPerPriorityFeeWei` for readability + maybe gas?

File: src/lib/PriorityOrderLib.sol [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    uint256 mpsPerPriorityFeeWei;
+    address recipient;
+}
+
+/// @dev External struct used to specify priority orders
+struct PriorityOrder {
+    // generic order information
+    OrderInfo info;
+    // the block at which the order becomes active
+    uint256 startBlock;
```
Comment (marktoda) [inline src/lib/PriorityOrderLib.sol:28]: I think this is good for now but one thing to consider for later --

this is signed over by the user, and users take unknown amt of sign to sign. In order to make sure this is in the future by the time the order is signed we will likely need to buffer quite a bit. May be difficult to tightly optimize for execution latency without some kind of cosigner situation just for updating the block

Comment (marktoda) [reply src/lib/PriorityOrderLib.sol:28]: potentially in the future users specify `minBlock` and `maxBlock` and labs can optionally cosign `targetBlock` with higher precision

File: test/lib/PriorityFeeLib.t.sol [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+        PriorityOutput memory output =
+            PriorityOutput({token: address(0), amount: 1 ether, mpsPerPriorityFeeWei: 100, recipient: address(0)});
+
+        OutputToken memory scaledOutput = PriorityFeeLib.scale(output, tx.gasprice);
+        uint256 scaledAmount = (output.amount * (MPS + tx.gasprice * output.mpsPerPriorityFeeWei)) / MPS;
+
+        assertEq(scaledOutput.amount, scaledAmount);
+    }
+
+    function testScaleInputPriorityFee_fuzz(uint256 priorityFee) public {
```
Comment (marktoda) [inline test/lib/PriorityFeeLib.t.sol:40]: some non-fuzz edge case tests would be good:
- input fee over max defaults to 0
- super high / low priority fee

**Review by marktoda — COMMENTED** [review]
Looks awesome!

File: src/lib/PriorityFeeLib.sol (hunk 11 above)
Comment (marktoda) [reply src/lib/PriorityFeeLib.sol:23]: oh sara already said this

File: src/lib/PriorityOrderLib.sol (hunk 4 above)
Comment (zhongeric) [reply src/lib/PriorityOrderLib.sol:102]: hmm this would require us to do `using PriorityOrderLib for PriorityOutput[]` within the same library? to get this syntax we would have to create a separate outputs library right

File: test/lib/PriorityFeeLib.t.sol (hunk 13 above)
Comment (zhongeric) [reply test/lib/PriorityFeeLib.t.sol:68]: I added explicit tests for when the mpsPerPriorityFeeWei is 0, when the amount is large enough to trigger overflow in the math library, for small priority fees, and for when input fee defaults to 0

**Review by marktoda — DISMISSED** [review]
LGTM!! :rocket:

**Review by marktoda — APPROVED** [review]
